### PR TITLE
Add curve aave strategy

### PR DESCRIPTION
**Pull request by olegnat**
Added have strategy + updated test infra to support customization.
Weekly prediction was mocked need to discuss if it worth to calculate it.

### CODE REVIEW

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../../../third_party/curve/ICurveDeposit3tokenUnderlying.sol";
+import "../../../third_party/curve/Gauge.sol";
+import "../StrategyBase.sol";
+
+
+/// @title Abstract contract for Wault strategy implementation
+/// @author belbix
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:24]: your name :)

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  string public constant STRATEGY_NAME = "CurveAaveStrategyFullBuyback";
+  /// @notice Version of the contract
+  /// @dev Should be incremented when contract changed
+  string public constant VERSION = "1.0.0";
+  /// @dev Placeholder, for non full buyback need to implement liquidation
+  uint256 private constant _BUY_BACK_RATIO = 10000;
+
+  /// @notice Wault rewards pool
+  address public pool;
+  /// @notice WexPolyMaster rewards pool ID
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:40]: is it curve ID? if yes change comments

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  uint256 public poolID;
+
+  /// @notice Contract constructor using on strategy implementation
+  /// @dev The implementation should check each parameter
+  /// @param _controller Controller address
+  /// @param _underlying Underlying token address
+  /// @param _vault SmartVault address that will provide liquidity
+  /// @param __rewardTokens Reward tokens that the strategy will farm
+  /// @param _pool WexPolyMaster pool address
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:49]: wrong comment

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  /// @return 0
+  function poolWeeklyRewardsAmount() external view override returns (uint256[] memory) {
+    uint256[] memory dummyResult = new uint256[](1);
+    return dummyResult;
+  }
+
+  /// @notice Hard to calculate for Curve.
+  /// @dev Don't use it in any internal logic, only for statistical purposes
+  /// @return 0
+  function computeWaultWeeklyPoolReward() public view returns (uint256) {
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:106]: you can remove this function

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  /// @dev Withdraw underlying and reward from Gauge pool
+  /// @param amount Deposit amount
+  function withdrawAndClaimFromPool(uint256 amount) internal override {
+    Gauge(pool).withdraw(amount, true);
+  }
+
+  /// @dev Exit from external project without caring about rewards
+  ///      For emergency cases only!
+  function emergencyWithdrawFromPool() internal override {
+    withdrawAndClaimFromPool(rewardPoolBalance());
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:137]: use withdraw without rewards
Gauge(pool).withdraw(amount, false);

File: contracts/third_party/curve/Gauge.sol [hunk 6]
```diff
@@ -0,0 +1,23 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity 0.8.4;
+
+interface Gauge {
+    function deposit(uint) external;
+    function balanceOf(address) external view returns (uint);
+    function claimable_reward(address _addr, address _token) external view returns (uint256);
+    function claimed_reward(address _addr, address _token) external view returns (uint256);
+    function claimable_reward_write(address _addr, address _token) external returns (uint256);
+    function withdraw(uint, bool) external;
```
Comment (belbix) [inline contracts/third_party/curve/Gauge.sol:10]: add attributes names, and better with comment from code
def withdraw(_value: uint256, _claim_rewards: bool = False):
    """
    @notice Withdraw `_value` LP tokens
    @dev Withdrawing also claims pending reward tokens
    @param _value Number of tokens to withdraw
    """

File: contracts/third_party/curve/Gauge.sol [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    function claim_rewards() external;
+}
+
+interface VotingEscrow {
```
Comment (belbix) [inline contracts/third_party/curve/Gauge.sol:14]: please, use separated files for each interface

File: contracts/infrastructure/PriceCalculator.sol [hunk 8]
```diff
@@ -67,6 +67,7 @@ contract PriceCalculator is Initializable, Controllable, IPriceCalculator {
   //In case of LP token, the underlying tokens will be found and valued to get the price.
   // Output token should exist int the keyTokenList
   function getPrice(address token, address outputToken) public view override returns (uint256) {
+
```
Comment (belbix) [inline contracts/infrastructure/PriceCalculator.sol:70]: remove

File: contracts/strategies/matic/curve/StrategyCurveAave.sol [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  address private constant CRV = address(0x172370d5Cd63279eFa6d502DAB29171933a610AF);
+
+  address[] private poolRewards = [WMATIC, CRV];
+  address[] private _assets;
+
+  constructor(
+    address _controller,
+    address _vault,
+    address _underlying,
+    uint256 _poolId
```
Comment (belbix) [inline contracts/strategies/matic/curve/StrategyCurveAave.sol:32]: I don't see where you use this poolId

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+  /// @notice Return approximately amount of reward tokens ready to claim in Aave Gauge pool
+  /// @dev Don't use it in any internal logic, only for statistical purposes. 
+  /// To get the result need to call claimable_reward_write from Gauge contract first, 
+  ///otherwise returns 0
+  /// @return Array with amounts ready to claim
+  function readyToClaim() external view override returns (uint256[] memory) {
+    uint256[] memory toClaim = new uint256[](_rewardTokens.length);
+    for (uint256 i = 0; i < _rewardTokens.length; i++) {
+      toClaim[i] = Gauge(pool).claimable_reward(address(this), _rewardTokens[i]);
```
Comment (belbix) [inline contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:83]: try to use claimable_reward_write as view

File: test/strategies/matic/curve/CurveStrategiesUniversalTest.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  await token.approve(depositerAddress, usdcUserBalance, {from: user.address});
+  
+  // swap USDC to underlying Curve.fi amDAI/amUSDC/amUSDT (am3CRV) (0xE7a24EF0C5e95Ffb0f6684b813A78F2a3AD7D171)
+  await depContract.add_liquidity([0, usdcUserBalance, 0], 0, true);
+}
+
+async function toClaimCalcFunc(strategy: StrategyBase) {
+  let gaugeAddress = "0x19793B454D3AfC7b454F206Ffe95aDE26cA6912c";
+  var gaugeContract = await ethers.getContractAt("Gauge", gaugeAddress) as Gauge;
+  await gaugeContract.claimable_reward_write(strategy.address, MaticAddresses.WMATIC_TOKEN);
```
Comment (belbix) [inline test/strategies/matic/curve/CurveStrategiesUniversalTest.ts:38]: what about CRV token?

File: test/strategies/matic/curve/CurveStrategiesUniversalTest.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  const token = await ethers.getContractAt("IERC20", MaticAddresses.USDC_TOKEN, user) as IERC20;
+  var usdcUserBalance = await Erc20Utils.balanceOf(MaticAddresses.USDC_TOKEN, user.address);
+  await token.approve(depositerAddress, usdcUserBalance, {from: user.address});
+  
+  // swap USDC to underlying Curve.fi amDAI/amUSDC/amUSDT (am3CRV) (0xE7a24EF0C5e95Ffb0f6684b813A78F2a3AD7D171)
+  await depContract.add_liquidity([0, usdcUserBalance, 0], 0, true);
+}
+
+async function toClaimCalcFunc(strategy: StrategyBase) {
+  let gaugeAddress = "0x19793B454D3AfC7b454F206Ffe95aDE26cA6912c";
```
Comment (belbix) [inline test/strategies/matic/curve/CurveStrategiesUniversalTest.ts:36]: use MticAddress

File: test/strategies/matic/curve/CurveStrategiesUniversalTest.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+describe('Universal Curve tests', async () => {
+  if (Settings.disableStrategyTests) {
+    return;
+  }
+
+    const idx = "1";
+    const lp_name = "CURVE_AAVE";
+    const underlying = "0xE7a24EF0C5e95Ffb0f6684b813A78F2a3AD7D171"; // am3CRV
```
Comment (belbix) [inline test/strategies/matic/curve/CurveStrategiesUniversalTest.ts:49]: use MaticAddress

File: test/strategies/matic/curve/CurveStrategiesUniversalTest.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    const token0_name = "am3CRV"; //Curve.fi amDAI/amUSDC/amUSDT (am3CRV)
```
Comment (belbix) [inline test/strategies/matic/curve/CurveStrategiesUniversalTest.ts:50]: use ERC20 symbol

File: test/strategies/matic/curve/CurveStrategiesUniversalTest.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+
+    console.log('strat', idx, lp_name);
+
+    startDefaultSingleTokenStrategyTest(
```
Comment (belbix) [inline test/strategies/matic/curve/CurveStrategiesUniversalTest.ts:55]: create a dedicated test file for curve, this logic pretty unique

File: test/strategies/DoHardWorkLoop.ts [hunk 16]
```diff
@@ -40,7 +40,7 @@ export class DoHardWorkLoop {
       }
       const loopStart = await StrategyTestUtils.getBlockTime();
       const balancesBefore = await StrategyTestUtils.saveBalances(info.signer.address, info.strategy);
-      // console.log('balancesBefore', balancesBefore[0].toString());
+      console.log('balancesBefore', balancesBefore[0].toString());
```
Comment (belbix) [inline test/strategies/DoHardWorkLoop.ts:43]: remove changes

File: test/strategies/DoHardWorkLoop.ts [hunk 17]
```diff
@@ -50,7 +50,7 @@ export class DoHardWorkLoop {
       await info.vault.doHardWork();
 
       const balancesAfter = await StrategyTestUtils.saveBalances(info.signer.address, info.strategy);
-      // console.log('balancesAfter', balancesAfter[0].toString(), balancesAfter[0].sub(balancesBefore[0]).toString());
+      console.log('balancesAfter', balancesAfter[0].toString(), balancesAfter[0].sub(balancesBefore[0]).toString());
```
Comment (belbix) [inline test/strategies/DoHardWorkLoop.ts:53]: remove changes

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol (hunk 1 above)
Comment (olegnat) [reply contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:24]: Ack

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol (hunk 2 above)
Comment (olegnat) [reply contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:40]: ack

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol (hunk 3 above)
Comment (olegnat) [reply contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:49]: ack

File: contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol (hunk 4 above)
Comment (olegnat) [reply contracts/base/strategies/curve/CurveAaveStrategyFullBuyback.sol:106]: ack

**Pull request closed without merging**